python/ray/data/block.py

```diff
@@ -512,21 +512,47 @@ def block_type(self) -> BlockType:
         raise NotImplementedError
 
 
-def _get_block_boundaries(columns: list[np.ndarray]) -> np.ndarray:
-    """Compute boundaries of the groups within a block, which is represented
-    by a list of 1D numpy arrays for each column. In each column,
-    NaNs/None are considered to be the same group.
+def _get_group_boundaries_sorted(
+    block: Block, key: Union[str, List[str]]
+) -> np.ndarray:
+    """
+    NOTE: THIS METHOD ASSUMES THAT PROVIDED BLOCK IS ALREADY SORTED
+
+    Compute boundaries of the groups within a block based on provided
+    key (a column or a list of columns)
+
+    NOTE: In each column, NaNs/None are considered to be the same group.
 
     Args:
-        columns: a list of 1D numpy arrays. This is generally given by the
-        dictionary values of ``BlockAccessor.to_numpy()``.
+        block: sorted block for which grouping of rows will be determined
+                based on provided key
+        key: column or list of columns determining the key for every
+                based on which the block will be grouped
 
     Returns:
         A list of starting indices of each group and an end index of the last
         group, i.e., there are ``num_groups + 1`` entries and the first and last
         entries are 0 and ``len(array)`` respectively.
     """
 
+    block_accessor = BlockAccessor.for_block(block)
+
+    # Get the list of boundaries including first start and last end indices
+    if key:
+        projected_block = block_accessor.to_numpy(key)
+
+        if isinstance(key, str):
+            projected_block = [projected_block]
+        else:
+            # projected_block is a dict of arrays
+            projected_block = list(projected_block.values())
+
+        return _get_group_boundaries_sorted_numpy(projected_block)
+
+    return np.array([0, block_accessor.num_rows()])
```

Comment:
why return this when key is empty? Should we just raise an error?

Reply:
It's semantically correct -- there's just 1 group. Otherwise, i'd need to handle that case separately at every caller